pkg/actor/upgrade.go

```diff
@@ -0,0 +1,192 @@
+package actor
+
+import (
+	"context"
+	"encoding/csv"
+	"fmt"
+	"github.com/Masterminds/semver/v3"
+	api "github.com/cockroachlabs/crdb-operator/api/v1alpha1"
+	"github.com/cockroachlabs/crdb-operator/pkg/condition"
+	"github.com/cockroachlabs/crdb-operator/pkg/kube"
+	"github.com/cockroachlabs/crdb-operator/pkg/resource"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubetypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+)
+
+func newUpgrade(scheme *runtime.Scheme, cl client.Client, config *rest.Config) Actor {
+	return &upgrade{
+		action: newAction("upgrade", scheme, cl),
+		config: config,
+	}
+}
+
+type upgrade struct {
+	action
+
+	config *rest.Config
+}
+
+func (up *upgrade) Handles(conds []api.ClusterCondition) bool {
+	return condition.False(api.NotInitializedCondition, conds)
+}
+
+func (up *upgrade) Act(ctx context.Context, cluster *resource.Cluster) error {
+	log := up.log.WithValues("CrdbCluster", cluster.ObjectKey())
+	log.Info("checking upgrade opportunities")
+
+	stsName := cluster.StatefulSetName()
+
+	key := kubetypes.NamespacedName{
+		Namespace: cluster.Namespace(),
+		Name:      stsName,
+	}
+	ss := &appsv1.StatefulSet{}
+	if err := up.client.Get(ctx, key, ss); err != nil {
+		return errors.Wrap(err, "failed to fetch statefulset")
+	}
+
+	if statefulSetIsUpdating(ss) {
+		return NotReadyErr{Err: errors.New("statefulset is updating, waiting for the update to finish")}
+	}
+
+	// Get version from spec
+	specStr := getVersionFromImage(cluster.Spec().Image)
+	if specStr == "" {
+		return fmt.Errorf("Unknown CockroachDB version in spec: %s", cluster.Spec().Image)
+	}
+
+	specVer, err := semver.NewVersion(specStr)
```

Comment:
We are actually considering moving away from deriving versions from a docker tag. An alternative approach we are considering is calling `cockroach version` to get the version. Anyway, this looks good for now, just sharing that we are discussing this with you.

Reply:
Noted and sounds good to me. It won't be hard to switch to this. Just create an issue or PR when the decision is made.